File: src/main/java/com/pttk/demo/service/SupplierService.java
package com.pttk.demo.service;

import com.pttk.demo.dto.request.SupplierCreationRequest;
import com.pttk.demo.dto.request.SupplierUpdateRequest;
import com.pttk.demo.exception.AppException;
import com.pttk.demo.exception.ErrorCode;
import com.pttk.demo.model.Supplier;
import com.pttk.demo.repository.SupplierRepository;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
public class SupplierService {
    SupplierRepository supplierRepository;

    public List<Supplier> getSuppliers() {
        return supplierRepository.findAll();
    }

    public Supplier getSupplier(String id) {
        return supplierRepository.findById(id)
                .orElseThrow(() -> new AppException(ErrorCode.SUPPLIER_NOT_EXISTED));
    }

    public Supplier createSupplier(SupplierCreationRequest request) {
        if (supplierRepository.existsByEmail(request.getEmail())) {
            throw new AppException(ErrorCode.USER_EXISTED);
        }

        Supplier supplier = new Supplier();
        supplier.setName(request.getName());
        supplier.setEmail(request.getEmail());
        supplier.setAddress(request.getAddress());
        supplier.setRepresentative(request.getRepresentative());
        supplier.setPhone(request.getPhone());

        return supplierRepository.save(supplier);
    }

    public Supplier updateSupplier(String id, SupplierUpdateRequest request) {
        Supplier supplier = getSupplier(id);

        if (request.getName() != null)
            supplier.setName(request.getName());
        if (request.getEmail() != null)
            supplier.setEmail(request.getEmail());
        if (request.getAddress() != null)
            supplier.setAddress(request.getAddress());
        if (request.getRepresentative() != null)
            supplier.setRepresentative(request.getRepresentative());
        if (request.getPhone() != null)
            supplier.setPhone(request.getPhone());

        return supplierRepository.save(supplier);
    }

    public void deleteSupplier(String id) {
        Supplier supplier = getSupplier(id);
        supplierRepository.delete(supplier);
    }

    public List<Supplier> searchSuppliers(String name) {
        return supplierRepository.findByNameContainingIgnoreCase(name);
    }
}

File: frontend/src/services/supplierService.js
import api from './api'
import { API_ENDPOINTS } from '../config/api'

export const supplierService = {
  getSuppliers: async () => {
    const response = await api.get(API_ENDPOINTS.SUPPLIERS.LIST)
    return response.data
  },

  getSupplier: async (id) => {
    const response = await api.get(API_ENDPOINTS.SUPPLIERS.GET(id))
    return response.data
  },

  createSupplier: async (supplierData) => {
    const response = await api.post(API_ENDPOINTS.SUPPLIERS.CREATE, supplierData)
    return response.data
  },

  updateSupplier: async (id, supplierData) => {
    const response = await api.put(API_ENDPOINTS.SUPPLIERS.UPDATE(id), supplierData)
    return response.data
  },

  deleteSupplier: async (id) => {
    const response = await api.delete(API_ENDPOINTS.SUPPLIERS.DELETE(id))
    return response.data
  },

  searchSuppliers: async (name) => {
    const response = await api.get(API_ENDPOINTS.SUPPLIERS.SEARCH, {
      params: { name },
    })
    return response.data
  },
}
